File: app/src/main/java/seigneur/gauvain/mycourt/ui/main/MainActivity.java
package seigneur.gauvain.mycourt.ui.main;

import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProvider;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.BottomNavigationView;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.view.MenuItem;
import android.widget.FrameLayout;
import android.widget.Toast;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import dagger.android.AndroidInjection;
import dagger.android.AndroidInjector;
import dagger.android.DispatchingAndroidInjector;
import dagger.android.support.HasSupportFragmentInjector;
import seigneur.gauvain.mycourt.R;
import seigneur.gauvain.mycourt.ui.base.BaseActivity;
import seigneur.gauvain.mycourt.ui.shotDraft.ShotDraftFragment;
import seigneur.gauvain.mycourt.ui.shotEdition.EditShotActivity;
import seigneur.gauvain.mycourt.ui.shots.ShotsFragment;
import seigneur.gauvain.mycourt.ui.user.UserFragment;
import seigneur.gauvain.mycourt.utils.FragmentStateManager;

public class MainActivity extends BaseActivity implements HasSupportFragmentInjector {

    @Inject
    ViewModelProvider.Factory viewModelFactory;

    private MainViewModel mMainViewModel;

    @BindView(R.id.fragment_place_holder)
    FrameLayout mFragmentContainer;

    @BindView(R.id.navigation)
    BottomNavigationView mNavigation;

    @BindView(R.id.fab_add_shot)
    FloatingActionButton mFabAddShot;

    private boolean isFAbVisible;

    private FragmentStateManager mFragmentStateManager;

    @Inject
    DispatchingAndroidInjector<Fragment> fragmentDispatchingAndroidInjector;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        AndroidInjection.inject(this);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);

        initFragmentManager(savedInstanceState);
        mNavigation.setOnNavigationItemSelectedListener(item -> {
            mMainViewModel.setmBottomNavPos(getNavPositionFromMenuItem(item));
            mMainViewModel.onBottomNavItemSelected();
            return true;
        });

        mNavigation.setOnNavigationItemReselectedListener(item -> {
            mMainViewModel.setmBottomNavPos(getNavPositionFromMenuItem(item));
            mMainViewModel.onBottomNavItemReselected();
        });

        mMainViewModel = ViewModelProviders.of(this, viewModelFactory).get(MainViewModel.class);
        mMainViewModel.init();

    }

    private void subscribeToSingleEvents(MainViewModel viewModel) {
        // BOTTOM NAV EVENTS
        viewModel.getNavItemSelectedEvent().observe(this, new Observer<Integer>() {
            @Override
            public void onChanged(Integer pos) {
                if (pos!=null && pos!=-1)
                    showFragment(pos);
            }
        });

        viewModel.getNavItemreselectedEvent().observe(this, new Observer<Integer>() {
            @Override
            public void onChanged(@Nullable Integer pos) {
                goBackOnTop(pos);
            }
        });

        viewModel.getbackNavSystemCommand().observe(
                this,
                pos -> {
                    goBackOnPrevItem(pos);
                }
        );

        viewModel.getFinishCommand().observe(this,
                call -> { finish(); }
                );

        viewModel.getEditCommand().observe(this,
                call -> { goToShotEdition(); }
        );
    }

    @Override
    public AndroidInjector<Fragment> supportFragmentInjector() {
        return fragmentDispatchingAndroidInjector;
    }

    @Override
    public void onResume() {
        super.onResume();
        subscribeToSingleEvents(mMainViewModel);
        mMainViewModel.checkIfTokenIsNull();

    }

    @Override
    public void onBackPressed() {
        //super.onBackPressed();
        mMainViewModel.onReturnNavigation();
    }

    @OnClick(R.id.fab_add_shot)
    public void goToEdition() {
        mMainViewModel.onAddFabclicked();
    }


    public void goBackAtStart(int position) {
        goBackOnTop(position);
    }


    public void closeActivity() {
        finish();
    }


    public void showFragment(int pos) {
        mFragmentStateManager.changeFragment(pos);
    }


    public void goBackOnPrevItem(int position) {
        mNavigation.setSelectedItemId(mNavigation.getMenu().getItem(position).getItemId());
    }


    public void goToShotEdition() {
        //todo - add some keys to intent - define from which  fragment user has clicked!
        Intent intent = new Intent(this, EditShotActivity.class);
        startActivity(intent);
    }


    public void showMessageShotPublished(){

    }


    public void showMessageShotDrafted() {

    }

    public void showNoInternetConnectionMessage(boolean showIt) {
        if(showIt) {
        }
    }

    public void showInternetConnectionRetrieved(boolean showIt) {
        if(showIt) {
        }

    }

    private void initFragmentManager(Bundle savedInstanceState) {
        showFAB(true);
        mFragmentStateManager = new FragmentStateManager(mFragmentContainer, getSupportFragmentManager()) {
            @Override
            public Fragment getItem(int position) {
                switch(position) {
                    case 0:
                        //return new ShotsFragment();
                        return new ShotsFragment();
                    case 1:
                        return new ShotDraftFragment();
                    case 2:
                        return new UserFragment();
                }
                //return new ShotsFragment();
                return new ShotsFragment();
            }
        };
        if (savedInstanceState == null) {
            mFragmentStateManager.changeFragment(0);
        }
    }

    int getNavPositionFromMenuItem(MenuItem menuItem) {
        switch (menuItem.getItemId()) {
            case R.id.navigation_home:
                showFAB(true);
                return 0;
            case R.id.navigation_dashboard:
                showFAB(true);
                return 1;
            case R.id.navigation_notifications:
                showFAB(false);
                return 2;
            default:
                return -1;
        }
    }

    public void goBackOnTop(int position) {
        Toast.makeText(this, "go back on top", Toast.LENGTH_SHORT).show();
    }

    public void showFAB(boolean show) {
        if (show && !isFAbVisible) {
            mFabAddShot.show();
            isFAbVisible=true;
        } else if (!show && isFAbVisible) {
            mFabAddShot.hide();
            isFAbVisible = false;
        }
    }

}

File: app/src/main/java/seigneur/gauvain/mycourt/ui/main/MainActivity.kt
package seigneur.gauvain.mycourt.ui.main

import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import android.content.Intent
import android.os.Bundle
import com.google.android.material.bottomappbar.BottomAppBar
import com.google.android.material.bottomnavigation.BottomNavigationView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import androidx.fragment.app.Fragment
import android.view.MenuItem
import android.view.View
import android.widget.FrameLayout
import android.widget.Toast

import javax.inject.Inject

import butterknife.BindView
import butterknife.ButterKnife
import butterknife.OnClick
import butterknife.Optional
import com.google.android.material.shape.MaterialShapeDrawable
import dagger.android.AndroidInjection
import dagger.android.AndroidInjector
import dagger.android.DispatchingAndroidInjector
import dagger.android.support.HasSupportFragmentInjector
import seigneur.gauvain.mycourt.R
import seigneur.gauvain.mycourt.ui.about.AboutFragment
import seigneur.gauvain.mycourt.ui.base.BaseActivity
import seigneur.gauvain.mycourt.ui.shotDraft.ShotDraftFragment
import seigneur.gauvain.mycourt.ui.shotEdition.EditShotActivity
import seigneur.gauvain.mycourt.ui.shots.ShotsFragment
import seigneur.gauvain.mycourt.ui.user.UserFragment
import seigneur.gauvain.mycourt.utils.Constants
import seigneur.gauvain.mycourt.utils.FragmentStateManager
import timber.log.Timber

class MainActivity : BaseActivity(), HasSupportFragmentInjector, CustomBottomActionMode.DraftListEditMode {

    @Inject
    lateinit var fragmentDispatchingAndroidInjector: DispatchingAndroidInjector<androidx.fragment.app.Fragment>

    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory

    private val mMainViewModel: MainViewModel by lazy {
         ViewModelProviders.of(this, viewModelFactory).get(MainViewModel::class.java)
    }

    @BindView(R.id.fragment_place_holder)
    lateinit var mFragmentContainer: FrameLayout

    @BindView(R.id.bottom_app_bar)
    lateinit var mBottomAppBar: BottomAppBar

    @BindView(R.id.navigation)
    lateinit var mNavigation: BottomNavigationView

    @BindView(R.id.fab_add_shot)
    lateinit var mFabAddShot: FloatingActionButton

    private var mFragmentStateManager: FragmentStateManager? = null

    var isEditModeActivated:Boolean?=false
    val mCustomBottomActionmode :CustomBottomActionMode by lazy {
        CustomBottomActionMode(this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        AndroidInjection.inject(this)
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        ButterKnife.bind(this)
        initFragmentManager(savedInstanceState)
        mMainViewModel.init()
        mNavigation.setOnNavigationItemSelectedListener { item ->
            mMainViewModel.setmBottomNavPos(getNavPositionFromMenuItem(item))
            mMainViewModel.onBottomNavItemSelected()
            true
        }

        mNavigation.setOnNavigationItemReselectedListener { item ->
            mMainViewModel.setmBottomNavPos(getNavPositionFromMenuItem(item))
            mMainViewModel.onBottomNavItemReselected()
        }

    }

    @Optional
    @OnClick(R.id.fab_add_shot)
    fun goToEdition() {
        if (isEditModeActivated!!)
            mMainViewModel.onDeleteClicked()
        else
            mMainViewModel.onAddFabclicked()
    }

    /*
    ************************************************************************************
    *   CustomBottomActionMode.DraftListEditMode
    ************************************************************************************/
    override fun onListEditStart() {
        mMainViewModel.changeEditMode(Constants.CUSTOM_ACTION_MODE_ON)
    }

    override fun onListEditStop() {
        mMainViewModel.changeEditMode(Constants.CUSTOM_ACTION_MODE_OFF)
    }

    /*
    ************************************************************************************
    *   PRIVATE FUNCTIONS
    ************************************************************************************/
    private fun subscribeToLiveData(viewModel: MainViewModel) {
        viewModel.actionModeEvent.observe(this,
                Observer { mode ->
                    if (mode==Constants.CUSTOM_ACTION_MODE_ON) {
                        setUpCustomActionMode(true)
                    } else {
                        setUpCustomActionMode(false)
                    }

                  })

    }

    private fun subscribeToSingleEvents(viewModel: MainViewModel) {
        // BOTTOM NAV EVENTS
        viewModel.navItemSelectedEvent.observe(this,
                Observer { pos ->
                    if (pos != null && pos != -1)
                        showFragment(pos)
        })

        viewModel.navItemreselectedEvent.observe(this, Observer { pos -> goBackOnTop(pos!!) })

        viewModel.getbackNavSystemCommand().observe(
                this,
                Observer<Int> {
                    pos -> goBackOnPrevItem(pos!!)
                }
        )

        viewModel.finishCommand.observe(this,
                Observer<Void> {
                    _ -> finish()
                }
        )

        viewModel.editCommand.observe(this,
                Observer<Void> {
                    _ -> goToShotEdition()
                }
        )
    }

    override fun supportFragmentInjector(): AndroidInjector<Fragment>? {
        return fragmentDispatchingAndroidInjector
    }

    public override fun onResume() {
        super.onResume()
        subscribeToSingleEvents(mMainViewModel)
        subscribeToLiveData(mMainViewModel)
        mMainViewModel.checkIfTokenIsNull()

    }

    override fun onBackPressed() {
        //super.onBackPressed()
        if (isEditModeActivated!!)
            mCustomBottomActionmode.stopMode()
        else
            mMainViewModel.onReturnNavigation()
    }

    fun goBackAtStart(position: Int) {
        goBackOnTop(position)
    }


    fun closeActivity() {
        finish()
    }


    private fun showFragment(pos: Int) {
        mFragmentStateManager?.changeFragment(pos)
    }


    private fun goBackOnPrevItem(position: Int) {
        mNavigation.selectedItemId = mNavigation.menu.getItem(position).itemId
    }


    private fun goToShotEdition() {
        val intent = Intent(this, EditShotActivity::class.java)
        startActivity(intent)
    }

    private fun initFragmentManager(savedInstanceState: Bundle?) {
        mFragmentStateManager = object : FragmentStateManager(mFragmentContainer, supportFragmentManager) {
            override fun getItem(position: Int): androidx.fragment.app.Fragment {
                when (position) {
                    0 -> return ShotsFragment()
                    1 -> return ShotDraftFragment()
                    3 -> return UserFragment()
                    4 -> return AboutFragment()
                }
                return ShotsFragment()
            }
        }
        if (savedInstanceState == null) {
            mFragmentStateManager!!.changeFragment(0)
        }
    }

    private fun getNavPositionFromMenuItem(menuItem: MenuItem): Int {
        when (menuItem.itemId) {
            R.id.navigation_home                -> return 0
            R.id.navigation_dashboard           -> return 1
            R.id.blank_nav_item                 -> return 2
            R.id.navigation_notifications       -> return 3
            R.id.navigation_about               -> return 4
            else                                -> return -1
        }
    }

    private fun goBackOnTop(position: Int) {
        Toast.makeText(this, "go back on top", Toast.LENGTH_SHORT).show()
    }

    private fun setUpCustomActionMode(isActivated:Boolean?=false) {
        isEditModeActivated=isActivated
        if (isActivated==true) {
            Timber.d("Edit Mode activated")
            mNavigation.visibility= View.INVISIBLE
            // Move FAB from the center of BottomAppBar to the end of it
            mBottomAppBar.fabAlignmentMode = BottomAppBar.FAB_ALIGNMENT_MODE_END
            // Change FAB icon
            mFabAddShot.setImageResource(R.drawable.ic_delete_black_24dp)
        } else {
            Timber.d("Edit Mode deactivated")
            mNavigation.visibility= View.VISIBLE
            // Move FAB from the center of BottomAppBar to the center of it
            mBottomAppBar.fabAlignmentMode = BottomAppBar.FAB_ALIGNMENT_MODE_CENTER
            // Change FAB icon
            mFabAddShot.setImageResource(R.drawable.ic_add_black_24dp)
        }
    }


}
